### ISSUE TITLE
Segfault in Flush of Particle Record Component

### ISSUE THREAD

**Issue by ax3l** [openPMD/openPMD-api#490]
**Describe the bug**
An early flush during parallel `CREATE` causes a segfault.

**To Reproduce**

Taken from `ParallelIOTest` section `hdf5_write_test` `[parallel][hdf5]` with an added `o.flush()` [here](https://github.com/openPMD/openPMD-api/blob/e3d816a615caba711c8ef85d82791a09387d72bd/test/ParallelIOTest.cpp#L101) after the first `storeChunk()`.

```C++
TEST_CASE( "hdf5_write_test", "[parallel][hdf5]" )
{
    int mpi_s{-1};
    int mpi_r{-1};
    MPI_Comm_size(MPI_COMM_WORLD, &mpi_s);
    MPI_Comm_rank(MPI_COMM_WORLD, &mpi_r);
    uint64_t mpi_size = static_cast<uint64_t>(mpi_s);
    uint64_t mpi_rank = static_cast<uint64_t>(mpi_r);
    Series o = Series("../samples/parallel_write.h5", AccessType::CREATE, MPI_COMM_WORLD);

    o.setAuthor("Parallel HDF5");
    ParticleSpecies& e = o.iterations[1].particles["e"];

    std::vector< double > position_global(mpi_size);
    double pos{0.};
    std::generate(position_global.begin(), position_global.end(), [&pos]{ return pos++; });
    std::shared_ptr< double > position_local(new double);
    *position_local = position_global[mpi_rank];

    e["position"]["x"].resetDataset(Dataset(determineDatatype(position_local), {mpi_size}));
    e["position"]["x"].storeChunk(position_local, {mpi_rank}, {1});

    o.flush();  //! @todo why does this flush cause a segfault?
}
```
**Expected behavior**
Should not segfault :-)

**Software Environment:**
 - version of openPMD-api: 0.8.0-dev (as of e3d816a615caba711c8ef85d82791a09387d72bd)
 - installed openPMD-api via: from source
 - operating system: Debian 9.7
 - machine: local laptop
 - name and version of Python implementation: N/A
 - version of HDF5: 1.10.0-patch1
 - version of ADIOS1: N/A
 - name and version of MPI: OpenMPI 2.0.2

**Comment by anokfireball** [comment]
Does serial execution of the same code also result in a segfault? If so, I might have a hunch (https://github.com/openPMD/openPMD-api/issues/321#issuecomment-417298702) about the origin.

**Comment by ax3l** [comment]
Not crashing, in serial mode we get the nice abort message:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  A Record can not be written without any contained RecordComponents: positionOffset
Aborted
```

**Comment by anokfireball** [comment]
Very weird that this results in a segfault in the parallel case.
For the actual problem: This is one of the few nasty edge cases of the standard.

We might want to disallow a flush before **both** `RecordComponents` from a `position` and `positionOffset` pair have been specified (so it does not irreversibly get flushed to disk with a undesired default). 
Alternatively, we might allow it, but then we will have to clean up when destructing the `Series` object (i.e. schedule missing `position` and `positionOffset` `RecordComponents` with defaults as per #321).

**Comment by ax3l** [comment]
Yes, something like the latter, potentially. As noted in [#321](https://github.com/openPMD/openPMD-api/issues/321#issuecomment-466415456), I think the sanest way is to clean up when destructing (or loosing user-reference) on the affected (particle) ~~`Record`~~ species.

By that, we can still auto-complete fine-grained with an open series and don't run in the same issue again while staging subsequent iterations later on.

**Comment by anokfireball** [comment]
> Do we know when a user destructs the handle of an `e` as in `auto e = o.iterations[1].particles["e"];` (and it's records)?

Ever since we merged https://github.com/openPMD/openPMD-api/pull/156, that is easily detectable. As everything is `std::shared_ptr`-based, we could simply check for `use_count == 1` (held by the API).
This would mean it's not too tricky, but does not solve the problem and is potentially dangerous. The user might temporarily drop the ref to a `ParticleSpecies` (or any handle for that matter) and then later re-take it from an upper node in the `Series` tree. If we create defaults based on that condition, the same irreversible and potentially undesired file state could be created.

I think the sanest and most helpful way is (as you proposed) to tie this to the lifetime of a `Series`. When that goes out of scope, the user has no way of triggering further flushes in the backend, meaning I/O is done (until the `Series` is manually reopened).
We than can decide whether we want
 - to throw (but only after the file has been closed in the backend so no state is lost) to indicate that the file is *NOT* respecting the standard
 - to 'auto-complete' the `Series` according to the standard

**Comment by ax3l** [comment]
Let's move the design discussion to #321 to keep things together and focus on the MPI-parallel bug fix here :) sorry for cross-posting!

**Comment by ax3l** [comment]
I can also reproduce a segfault with `examples/9_particle_write_serial.py`, but did not check if it's just an uncatched `throw` or the same issue.
